### ISSUE TITLE
How can I create a save view into external memory through `TypeData`?

### ISSUE THREAD

**Issue by blaugold** [dart-lang/sdk#48704]
I have some externally allocated memory and would like to allow code to access it without having to know about the lifetime of the memory. I can get a `Uint8List` through `Pointer.asTypedList` but I have to manually manage the lifetime of the backing memory.

Is it safe to attach a native finalizer to the object returned from `Pointer.asTypedList`, which frees the external memory, through `Dart_NewFinalizableHandle`? 
Also, what would be the implications for the `Uint8List`'s `ByteBuffer`? I suspect it would not be safe to use it.

**Comment by devoncarew** [comment]
Perhaps @dcharkes would know this? Not sure if this relates to dart:ffi or not.

**Comment by dcharkes** [comment]
> Is it safe to attach a native finalizer to the object returned from Pointer.asTypedList, which frees the external memory, through Dart_NewFinalizableHandle?

Yes, but you need to take care to hold on to the object which has the finalizer. Indeed, if you use the bytebuffer to make a new typed data and lose the old one, the finalizer could run.

Also, to ensure objects don't get collected prematurely you can add a Dart method to keeps it alive.

```
pragma('vm:never-inline')
void reachabilityFence(Object? object) {}
```

In Dart 2.17 we make using native finalizers much easier.

First, we make them available in Dart code, rather than using the Dart API.

https://github.com/dart-lang/sdk/blob/532c116cd2db02adcb6dbf32f19d5c438fd3585b/sdk/lib/ffi/native_finalizer.dart#L232-L297

Note that this only works for `void function(void*)`, including `free` but excluding `WinHeapFree`. So in Dart 2.17 this doesn't work for Windows yet. Supporting Windows is tracked in:

* https://github.com/dart-lang/sdk/issues/47778

Secondly, we introduce `Finalizable`, which prevents the need manually writing `reachabilityFence` by automatically extending the lifetime of variables to the end of the syntactic scope.

https://github.com/dart-lang/sdk/blob/532c116cd2db02adcb6dbf32f19d5c438fd3585b/sdk/lib/ffi/native_finalizer.dart#L7-L223

And thirdly, we standardize on using a wrapper class wrapping the `Pointer`, `TypedData` or `ByteBuffer` (or file handle, etc.) and standardize on always using the native resource as methods of that class rather than ever leaking the resource form out of the class. An example can be found here:

https://github.com/dart-lang/sdk/blob/532c116cd2db02adcb6dbf32f19d5c438fd3585b/tests/ffi/vmspecific_native_finalizer_test.dart#L40-L71

If you're using Dart 2.16 you can adopt the same patterns by wrapping your `TypedData` or `Pointer` in a wrapper class, attaching the finalizer with `Dart_NewFinalizableHandle` to the wrapper class, only using the typed data in the wrapper class, and adding a call to `reachabilityFence(this)` at the end of each method in the wrapper class.

It should be fairly easy to migrate this to the `NativeFinalizer` with `Finalizable` in Dart 2.17 (if you don't have to support Windows) or in Dart 2.18 then.

**Comment by blaugold** [comment]
Thanks for the detailed response!

In my use cases I have another API which takes a `Uint8List` as an argument and I cannot easily control for how long this list will be used. Currently, I copy the external data into a new `Uint8List`, but was hoping to be able to remove the extra allocation and copying.

As an aside, is there a difference between these to reachability fences?
```dart
@pragma('vm:never-inline')
void reachabilityFence(Object? object) {}
```
```dart
@pragma('vm:never-inline')
Object? reachabilityFence(Object? obj) {
  return obj;
}
```

I was using the first variant for the longest time but got crashes in AOT mode which looked like being caused by premature finalization of the native resource.

**Comment by dcharkes** [comment]
> I was using the first variant for the longest time but got crashes in AOT mode which looked like being caused by premature finalization of the native resource. I'm not seeing those crashes with the second variant.

That is very curious. I would have to investigate that.

> In my use cases I have another API which takes a `Uint8List` as an argument and I cannot easily control for how long this list will be used. Currently, I copy the external data into a new `Uint8List`, but was hoping to be able to remove the extra allocation and copying.

Unfortunately that is currently not supported. I'll give it some thought. Maybe it would be possible to convert a `Pointer` into a `TypedData` that uses the VM's typed data finalizer instead of a manual `Dart_NewFinalizableHandle` or `NativeFinalizer`. In other words expose `Dart_NewExternalTypedDataWithFinalizer` as a method on `Pointer`.

**Comment by dcharkes** [comment]
> Also, what would be the implications for the `Uint8List`'s `ByteBuffer`? I suspect it would not be safe to use it.

I've dug some deeper on this: And it looks like it is actually safe to use the buffers.

The buffer wraps the typed data: 

https://github.com/dart-lang/sdk/blob/6faa5f3bd00ad8cbc640b3fc80cf7466c002a7df/sdk/lib/_internal/vm/lib/typed_data_patch.dart#L2042

And getting another TypedData creates a view around the same backing store:

https://github.com/dart-lang/sdk/blob/6faa5f3bd00ad8cbc640b3fc80cf7466c002a7df/sdk/lib/_internal/vm/lib/typed_data_patch.dart#L1914-L1920

The `Dart_NewExternalTypedDataWithFinalizer` simply allocates an external typed data (which is the same object for an `asTypedList`) and then attaches the finalizer:

https://github.com/dart-lang/sdk/blob/6faa5f3bd00ad8cbc640b3fc80cf7466c002a7df/runtime/vm/dart_api_impl.cc#L3808-L3813

So, doing `asTypedList` in Dart, and then `Dart_NewFinalizableHandle` in C/C++ _or_ `NativeFinalizer` in Dart should be equal to using `Dart_NewExternalTypedDataWithFinalizer` in C/C++.
